File: backend/nodejs/src/models/Schedule.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../database/connection');

const Schedule = sequelize.define('Schedule', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true
  },
  route_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'routes',
      key: 'id'
    }
  },
  van_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'vans',
      key: 'id'
    }
  },
  departure_date: {
    type: DataTypes.DATEONLY,
    allowNull: false
  },
  departure_time: {
    type: DataTypes.TIME,
    allowNull: false
  },
  price: {
    type: DataTypes.DECIMAL(10, 2),
    allowNull: false
  },
  available_seats: {
    type: DataTypes.INTEGER,
    defaultValue: 12,
    allowNull: false
  },
  status: {
    type: DataTypes.ENUM('scheduled', 'departed', 'completed', 'cancelled'),
    defaultValue: 'scheduled',
    allowNull: false
  }
}, {
  tableName: 'schedules',
  timestamps: true,
  underscored: true,
  indexes: [
    {
      fields: ['departure_date', 'route_id']
    }
  ]
});

module.exports = Schedule;

File: backend/golang/internal/model/schedule.go
package model

import "time"

// Schedule represents a scheduled trip (ตาม database schema ที่แท้จริง)
type Schedule struct {
	ID             int       `json:"id"`
	RouteID        int       `json:"route_id"`
	VanID          int       `json:"van_id"`
	DepartureTime  time.Time `json:"departure_time"` // TIMESTAMP in database
	ArrivalTime    time.Time `json:"arrival_time"`   // TIMESTAMP in database
	Price          float64   `json:"price"`
	AvailableSeats int       `json:"available_seats"`
	Status         string    `json:"status"` // "active", "cancelled", "completed"
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

// ScheduleWithDetails schedule พร้อมข้อมูล route และ van (สำหรับ JOIN queries)
type ScheduleWithDetails struct {
	ID             int       `json:"id"`
	RouteID        int       `json:"route_id"`
	VanID          int       `json:"van_id"`
	DepartureTime  time.Time `json:"departure_time"`
	ArrivalTime    time.Time `json:"arrival_time"`
	Price          float64   `json:"price"`
	AvailableSeats int       `json:"available_seats"`
	Status         string    `json:"status"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
	
	// Joined fields from routes table
	Origin          string  `json:"origin"`
	Destination     string  `json:"destination"`
	DurationMinutes int     `json:"duration_minutes"`
	DistanceKM      float64 `json:"distance_km"`
	
	// Joined fields from vans table
	VanNumber    string `json:"van_number"`
	LicensePlate string `json:"license_plate"`
	Driver       string `json:"driver"`
	TotalSeats   int    `json:"total_seats"`
}

// CreateScheduleRequest สำหรับสร้างตารางเวลา
type CreateScheduleRequest struct {
	RouteID        int    `json:"route_id" binding:"required"`
	VanID          int    `json:"van_id" binding:"required"`
	DepartureTime  string `json:"departure_time" binding:"required"` // ISO 8601 format
	ArrivalTime    string `json:"arrival_time" binding:"required"`   // ISO 8601 format
	Price          float64 `json:"price" binding:"required"`
	AvailableSeats int    `json:"available_seats" binding:"required"`
	Status         string `json:"status" binding:"required,oneof=active cancelled completed"`
}

// UpdateScheduleRequest สำหรับแก้ไขตารางเวลา
type UpdateScheduleRequest struct {
	RouteID        int     `json:"route_id"`
	VanID          int     `json:"van_id"`
	DepartureTime  string  `json:"departure_time"`
	ArrivalTime    string  `json:"arrival_time"`
	Price          float64 `json:"price"`
	AvailableSeats int     `json:"available_seats"`
	Status         string  `json:"status" binding:"omitempty,oneof=active cancelled completed"`
}
